shell/browser/api/electron_api_session.cc

```diff
@@ -149,6 +152,41 @@ uint32_t GetQuotaMask(const std::vector<std::string>& quota_types) {
   return quota_mask;
 }
 
+constexpr content::BrowsingDataRemover::DataType kClearDataTypeAll = ~0ULL;
+constexpr content::BrowsingDataRemover::OriginType kClearOriginTypeAll =
+    content::BrowsingDataRemover::ORIGIN_TYPE_UNPROTECTED_WEB |
+    content::BrowsingDataRemover::ORIGIN_TYPE_PROTECTED_WEB;
+
+// Observes the BrowsingDataRemover that backs the `clearBrowsingData` method
+// and resolves/rejects that API's promise once it's done. This type manages its
+// own lifetime, deleting itself once it's done.
+class ClearBrowsingDataObserver
+    : public content::BrowsingDataRemover::Observer {
+ public:
+  ClearBrowsingDataObserver(gin_helper::Promise<void> promise,
+                            content::BrowsingDataRemover* remover)
+      : promise_(std::move(promise)) {
+    observation_.Observe(remover);
+  }
+
+  void OnBrowsingDataRemoverDone(
+      content::BrowsingDataRemover::DataType failed_data_types) override {
+    if (failed_data_types == 0ULL) {
+      promise_.Resolve();
+    } else {
+      promise_.RejectWithErrorMessage(base::StringPrintf(
+          "Failed to clear browsing data (%" PRIu64 ")", failed_data_types));
```

Comment:
It looks like `v8::Error` can be cast to an object to set additional properties [according to the v8 tests.](https://source.chromium.org/chromium/chromium/src/+/main:v8/test/cctest/test-api.cc;l=4772-4776;drc=94f819f39ee5b698e0b82cbcefb2d6cdae640913)

Maybe the failed data types could be set as a custom property on the error object. It would be a little unusual though as Electron doesn't do this anywhere else.

Reply:
I had that behavior in a previous revision, and then I realized that it would be better if the failed data types was a set or array of data categories as strings. It would be far more useful to JS code in that form. That mapping doesn't really exist right now but might in a future revision.

Reply:
I think there are two approaches to pick from:
1. Keep this error as-is for now, and extend it in the future when the mapping from data types to category strings exists.
2. Same as (1) where we add a simple field in the future, but attach an `errorCode` field now with the raw numerical value of `failed_data_types`.

There's some rough edges with exposing this raw value to JS though. It's a 64-bit unsigned integer, which could theoretically go beyond the 53-bit limit of normal JS numbers, though there are barely half that number of flags right now. The safe route would be to expose it as a bigint, but that could be confusing.